### ISSUE TITLE
sql,server: statements and combined statements endpoints errors

### ISSUE THREAD

**Issue by xinhaoz** [cockroachdb/cockroach#71245]
To date we have observed several errors from the /statments and /combined-statements endpoints.

## Issue 1: /Statements response too large for gRPC
Currently, /Statements, returns all cluster-wide in-memory statement stats.
Since the virtual statement tables call /Statements, this is also an issue for the combined stats API.
The virtual tables calling /Statement are particularly wasteful as only one of the `statements` or `transactions` portion of the response is used, and the other part discarded. We should at least work to remove the unnecessary collection and return of statement / transaction stats when not necessary ([see this issue for more context](https://github.com/cockroachdb/cockroach/issues/71829)).

We have seen reported issues where this might be the cause:  https://github.com/cockroachlabs/support/issues/1263.

Potential solutions:
1. gRPC streaming: If we could stream the statement stats from each node, and perform an ordered merge of those results, we could avoid having to buffer the entire cluster-wide stmt stats results. However, we would need to open connections to all nodes concurrently, this may be an issue on large clusters with many nodes. Not too sure on what our limitations here are for maximum number of connections, noted the [max number of requests are currently capped at 100](https://github.com/cockroachdb/cockroach/blob/ad4ba0a483b8c2d3c8d8af667258e86628de6faf/pkg/server/status.go#L96)
2. Aggregate in-memory statement stats on nodeId: this could reduce the size of the response considerably if many nodes share the same statement keys. This would require more investigation to see if the nodeId field in each individual statement stat is still necessary.

## Issue 2: Full scans on system stats tables might be causing timeouts
Another issue we are seeing are timeouts. One cause for this error may be due to the fact that the virtual stats tables (`crdb_internal.statement_statistics` and `crdb_internal.transaction_statistics`)  performs a full-scan on the system stats tables (`system.statement_statistics` and `system.transaction_statistics`). After we have been flushing stats for a while, these tables can grow large enough that a full scan might be causing timeouts to occur.

### Possible solution to both problems
@Azhng proposed a solution that might address both of the problems above.
In our sqlstats subsystem, we currently flush in-memory stats to the system tables on an hourly interval, or when they grow too large. Instead of buffering cluster-wide in-memory sql stats and merging them with the system tables, we can:

1.  Perform a flush on every request to retrieve combined stats, thus system tables now have the most up to date stats at time of request
2.  Directly query from the system tables with the necessary filters
3. Paginate results (can be now be done without worrying about in-memory stats changing)

This avoids the full table scans, and the buffering of in-memory stats, removing the necessity for the virtual tables entirely. Some performance testing is necessary here, as we'd be flushing much more often here than the current hourly interval.

Epic CRDB-11997

gz#10996

**Comment by xinhaoz** [comment]
We'll be working on this in the next milestone (this month).

**Comment by Azhng** [comment]
Some updates after the gRPC streaming idea bankrupted: 

I ran the TPCC-2500 on a 3-node cluster with `set cluster setting sql.stats.flush.interva = '1s'`. (1 second flush interval)

The result is very good: 

| Branch                               | tpmC    | %efficiency | Avg (ms) | p50 (ms) | p75 (ms) | p90 (ms) | p99 (ms) | pMax (ms) |
|--------------------------------------|---------|-------------|----------|----------|----------|----------|----------|-----------|
| master                               | 31386.0 | 97.6%       |  76.2    |  30.4    | 58.7     |  71.3    | 109.1    | 23622.3   |
| master                               | 31430.2 | 97.8%       |  50.4    |  29.4    | 58.7     |  71.3    | 113.2    | 11811.2   |
| 1s-flush                             | 31417.4 | 97.7%       |  56.6    |  31.5    | 62.9     |  75.5    | 117.4    | 16106.1   |


We barely see any latency drop nor any drop in efficiency. The delta here is well within the range of statistical uncertainty. 
This leads me to think that we have a few options at our hand right now:

---

# Alternative 1 

The benchmark result leads me to think that our previous 1-hour flush interval is way too conservative and perhaps we don't even need to have a virtual table at all. Just flush SQL Stats into system table at a high frequency (doesn't need to be 1 second, perhaps 15 seconds) and we can just directly query the system table. 

Moreover, our current implementation of the flusher is not the most optimal, if we implement batching (#67867), we can make the flusher even more efficient. 

This also solved our problem with suboptimal query plan identified in #74680. Nevertheless, we should test this in a relatively larger cluster. (perhaps 10 to 20 or maybe 30 nodes?) 

However, note that implementing this solution means our data cardinality is going to increase by a significant amount. (aggregated by 10 seconds vs  aggregated by 1 hour). This is because our current aggregation interval is coupled with flush interval. This means that we should probably take on #69302 to implement the compaction within the SQL Stats cleanup job. This could be quite a lot of work. I'm eyeballing about 2 ~ 4 weeks.

--- 

# Alternative 2 

Alternative 2 is a variant of alternative 1, we can decouple the flush interval with aggregation interval, where we flush at a 10-second interval, but the flusher still aggregates within the 1 hour bucket controlled via another cluster setting. This makes the data cardinality problem mentioned in the alternative 1, without the **immediate** need to implement compaction. (Though personally I think we should still take that on).

--- 

# Alternative 3

Instead of modifying the write-path. The read-path in `/Statement` endpoint simply makes a cluster RPC fanout to instruct all nodes in cluster to flush their SQL stats into the system table, then use a follower read to read the system table. This is the simplest (probably doable within a week) but also the hackiest approach. In the big cluster, I'm afraid we might run into thundering herd problem by doing this. 


---

# Bottom Line

All three alternatives here moves away from using the virtual table to directly using the system table. This allows us to perform better scans and be more efficient. Moreover, by using the system table, we can use use MVCC read to have a consistent view of the system table so we can easily implement pagination without worrying about data consistency, which also address the first problem listed in the issue.

cc: @maryliag

**Comment by Azhng** [comment]
# Quick follow up 1:

The time estimation above was mainly for the backend change. The frontend change is unaccounted for.

# Quick follow up 2:

All three alternatives presented above all opt to use the system table over virtual table. This is really due to the inherit limitation within the virtual table not be able to pass down the scan constraints. A fourth alternative would just be fix the virtual table to be aware of the scan constraints. But I'm not too familiar with virtual table so this alternative would be the one with the most of unknown unknowns.

**Comment by xinhaoz** [comment]
I am really tempted to go with the approach that kills the virtual tables. With that approach, I am still advocating for aggregating on nodeIds after querying the system table to reduce response size at the cost of by node stats (we can still preserve which nodes the stmt was executed on). Pagination might be a better long term solution, but we likely need more time to discuss/flesh that out in terms of how that affects frontend tables loading, how much we need to modify our requests/response msgs etc..

**Comment by Azhng** [comment]
Just for more context, the `node_id` field was really not meant for data storage when the persisted sql stats was designed, it was served as a way to shard the write traffic to prevent contention. Original idea was to have a compaction job to aggregate sql stats across `node_id` boundaries. :shrug:

**Comment by maryliag** [comment]
I like the idea of using the system table directly, without the need to combine in-memory and persisted. 
From those alternatives, I prefer 1 & 2 instead of 3, because waiting for the flush on every read can have a delay, meaning having it ready would get faster results.

Regarding 1 vs 2: on option 1 you have more granular data and your compaction is bringing them to 1h (for example), and the option 2 is adding them already by the hour, creating the same result at the end, but option 1 having extra steps and possible causing issues with the amount of data, so I suggesting going with option 2 (at least until we have any feedback from customers requiring a more granular aggregation)

**Comment by Azhng** [comment]
Just realized another thing related to high frequency flush. It's not necessary an immediate issue or blocker, but something we should definitely keep in mind.

The alternative 2 implies that we will be running a lot of `UPDATE` statements. This might not change the data cardinality much within the SQL table, but we will be creating a lot of MVCC garbage in the background. With default Row GC TTL=25 hours right now, we will probably be paying a hefty price for all those MVCC garbage during scan. How much of an impact will this be? I'm not sure. We can probably find out about this by running a cluster for longer than 25 hours and see what happens. But it's probably still going to be way faster than our current full table scan + disk spilling sort.

The upcoming row-level TTL project should make this problem a lot better. Though I'm not quite sure with the timeline with that project.

**Comment by ajwerner** [comment]
I feel like we need to attack the root cause of the problem here head-on: as we had implemented the virtual table, we couldn't efficiently combine the persistent data with the in-memory data. Fortunately we already have the technology to solve that problem. See https://github.com/cockroachdb/cockroach/pull/74778.

**Comment by ajwerner** [comment]
> The upcoming row-level TTL project should make this problem a lot better. Though I'm not quite sure with the timeline with that project.

I don't think this does what you think it does. It will not in any way help with garbage collection of expired rows. It's only to delete rows which nothing else deleted.

**Comment by maryliag** [comment]
re-opening since is not yet backported

**Comment by maryliag** [comment]
issue backported
